### ISSUE TITLE
Use locals() to capture 'plan_args' metadata.

### ISSUE THREAD

**Issue by danielballan** [bluesky/bluesky#1304]
In the built-in plans and in high-quality user-written plans, we use and encourage the pattern:

```py
def my_plan(detectors, a, b, c=some_default, md=None):

    _md = {
        'detectors': [det.name for det in detectors],
        'plan_args': {'detectors': list(map(repr, detectors)), 'a': a, 'b': b, 'c': c},
        'plan_name': 'my_plan',
    }

    # If a key exists in md, it overwrites the default in _md.
    _md.update(md or {})
    ...
```

where a, b, c are stand-ins for whatever a plan’s particular parameters may be. It is easy to forget to update the plan_args metadata when a new parameter is added.

Michael Sutherland pointed out that we could spell it like this instead:

```py
def my_plan(detectors, a, b, c=some_default, md=None):

    _md = {
        'detectors': [det.name for det in detectors],
        'plan_args': dict(locals()),
        'plan_name': 'my_plan',
    }
    # We can’t put Device objects in md. Replace them with strings.
    _md['plan_args']['detectors'] = list(map(repr, detectors))

    # If a key exists in md, it overwrites the default in _md.
    _md.update(md or {})
    ...
```

Here, `locals()` does most of the work capturing the names of the plan’s arguments and their values. Of course, it’s important to run it at the top of the function body before any new values are added to the namespace or any of the input variables are modified. And we have to do a little sanitization for any variables that aren’t JSON- and Mongo-serializable, but that can be done after the fact as shown.

I’m not sure I would call this code _simpler_ but I might call it easier to maintain.

**Comment by prjemian** [comment]
Is the *encouraged* pattern in the documents?  It should be.

**Comment by danielballan** [comment]
Yes, the example above is lightly adaptive from one in this section of the tutorial: https://blueskyproject.io/bluesky/tutorial.html#add-metadata
